src/managers/backup.py

```diff
@@ -176,6 +183,55 @@ def format_backups_table(
 
         return out_f.getvalue()
 
+    def is_snapshot_in_bucket(self, backup_id: str) -> bool:
+        """Check whether the requested snapshot to restore is in the object storage."""
+        try:
+            bucket = self.bucket
+        except KeyError:
+            return False
```

Comment:
**question**: The pattern we use to check for all possible failures in the action event handler is fine, but sometimes we need some out-of-context error catching (here, for instance, logically we would already have checked for a missing s3 integrator, but the failure evaluation is eager, not lazy). For future reference, do you prefer to:
- keep those errors catching like this, with a comment specifying why we have this check
- make failures lazy evaluated by wrapping them up in a lambda, so that they are not executed if a failure condition has already been found true
- change the pattern on the action event handler side from "list all possible failures, report the first one" to "check one at a time"

Reply:
If I have to suggest my preference, that would be to make them lazy, using a lambda possibly with also a 'context' argument. 

This way one could even write a standalone condition as:

```
lambda context: BackupManager(context.state).is_snapshot_in_bucket(bucket_id)
```

and one could also compose conditions, that feels very functional

But up to you guys, this is definitely your call!

Reply:
> Could we make that property type Bucket | None

TBH I don't really like this approach because it then requires checking if `self.bucket is None` everywhere, but this is already taken care of by the preliminary checks in the actions

Reply:
Just to make sure everyone is one the same page, the issue I am describing is having the pattern
```python
failures = [
    condition1,
    condition2,
    condition3,
]
for failure in failures:
    if failure:
         ...
```
If `condition2` is a prerequisite for `condition3`, this can err in "weird" ways. I say "weird" because from a functional PoV we do check `condition2` before `condition3`, but since the evaluation in the list is eager, now we have to include some defensive programming in `condition3`.

What Enrico and I seem to like is to make the conditions lazy by wrapping them up in a lambda for instance, so that we end up with:
```python
failures = [
    lambda: condition1,
    lambda: condition2,
    lambda: condition3,
]
for failure in failures:
    if failure():  # Only evaluated at this point
         ...
```

Reply:
Changed in 653db76a65a4